File: zmusic-plugin/src/main/java/me/zhenxin/zmusic/api/bossbar/BossBarBukkit.java
package me.zhenxin.zmusic.api.bossbar;

import me.zhenxin.zmusic.utils.runtask.BukkitTaskScheduler;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

import java.util.List;


public class BossBarBukkit implements BossBar {

    private final Player p;
    private volatile String title;
    private final double seconds;
    private final BarColor color;
    private final BarStyle style;
    private org.bukkit.boss.BossBar bar;
    private Runnable cancelProgressTask;

    public BossBarBukkit(Object p, String title, BarColor color, BarStyle style, float seconds) {
        Player player = (Player) p;
        this.p = player;
        this.title = title;
        this.color = color;
        this.style = style;
        this.seconds = seconds;
    }

    @Override
    public void showTitle() {
        BukkitTaskScheduler.run(p, this::showTitleSync);
    }

    private void showTitleSync() {
        if (bar == null) {
            bar = Bukkit.createBossBar(title, org.bukkit.boss.BarColor.valueOf(color.name()),
                    org.bukkit.boss.BarStyle.valueOf(style.name()));
        }
        bar.setVisible(true);
        bar.setProgress(0);
        bar.addPlayer(p);
        Runnable tick = () -> {
            if (!bar.isVisible()) {
                cancelTask();
                return;
            }
            double step = 1F / seconds;
            double prog = Math.min(bar.getProgress() + step, 1.0);
            if (prog >= 1.0) {
                bar.setProgress(1.0);
                bar.setVisible(false);
                cancelTask();
                return;
            }
            bar.setProgress(prog);
        };
        cancelProgressTask = BukkitTaskScheduler.runAtFixedRate(p, tick, 20L, 20L);
    }

    @Override
    public String getTitle() {
        return title;
    }

    @Override
    public void setTitle(String title) {
        this.title = title;
        BukkitTaskScheduler.run(p, () -> {
            if (bar != null) {
                bar.setTitle(title);
            }
        });
    }

    @Override
    public void removePlayer(Object player) {
        Player p = (Player) player;
        BukkitTaskScheduler.run(p, () -> {
            if (bar != null) {
                bar.removePlayer(p);
            }
            cancelTask();
        });
    }

    @Override
    public void removeAll() {
        BukkitTaskScheduler.run(p, () -> {
            if (bar != null) {
                bar.removeAll();
            }
            cancelTask();
        });
    }

    @Override
    public boolean isVisible() {
        return bar != null && bar.isVisible();
    }

    @Override
    public void setVisible(boolean visible) {
        BukkitTaskScheduler.run(p, () -> {
            if (bar != null) {
                bar.setVisible(visible);
            }
            if (!visible) {
                cancelTask();
            }
        });
    }

    private void cancelTask() {
        if (cancelProgressTask != null) {
            cancelProgressTask.run();
            cancelProgressTask = null;
        }
    }

    @Override
    public BarColor getBarColor() {
        return BarColor.valueOf(bar.getColor().name());
    }

    @Override
    public void setBarColor(BarColor barColor) {
        bar.setColor(org.bukkit.boss.BarColor.valueOf(barColor.name()));
    }

    @Override
    public BarStyle getBarStyle() {
        return BarStyle.valueOf(bar.getStyle().name());
    }

    @Override
    public void setBarStyle(BarStyle barStyle) {
        bar.setStyle(org.bukkit.boss.BarStyle.valueOf(barStyle.name()));
    }

    @Override
    public double getProgress() {
        return (float) bar.getProgress();
    }

    @Override
    public void setProgress(double progress) {
        BukkitTaskScheduler.run(p, () -> {
            if (bar != null) {
                bar.setProgress(progress);
            }
        });
    }

    @Override
    public boolean hasPlayer(Object playerObj) {
        List<Player> players = bar.getPlayers();
        for (Player player : players) {
            if (player == playerObj)
                return true;
        }
        return false;
    }

    @Override
    public void addPlayer(Object playerObj) {
        Player player = (Player) playerObj;
        BukkitTaskScheduler.run(player, () -> {
            if (bar != null) {
                bar.addPlayer(player);
            }
        });
    }
}

File: zmusic-plugin/src/main/kotlin/me/zhenxin/zmusic/module/bossbar/impl/BossBarBukkit.kt
package me.zhenxin.zmusic.module.bossbar.impl

import me.zhenxin.zmusic.module.bossbar.BossBar
import org.bukkit.boss.BarColor
import org.bukkit.boss.BarStyle
import taboolib.common.platform.Platform
import taboolib.common.platform.PlatformImplementation
import taboolib.common.platform.ProxyPlayer
import taboolib.common.platform.function.submit
import taboolib.platform.BukkitPlugin

/**
 * BossBar Bukkit实现
 *
 * @author 真心
 * @since 2022/7/5 14:05
 * @email qgzhenxin@qq.com
 */
@PlatformImplementation(Platform.BUKKIT)
class BossBarBukkit(player: ProxyPlayer) : BossBar(player) {
    private val bukkitPlugin = BukkitPlugin.getInstance()

    private val bar = bukkitPlugin.server.createBossBar("", BarColor.BLUE, BarStyle.SEGMENTED_20)
    private var seconds = 0F

    override fun start() {
        if (player.isOnline()) {
            bar.isVisible = true
            bar.progress = 0.0
            bar.addPlayer(player.cast())
            submit(async = true) {
                val step = 1F / seconds
                var progress = bar.progress
                while (progress >= 0 || progress <= 1) {
                    progress += step
                    if (progress > 1) break
                    bar.progress = progress
                    if (bar.isVisible) {
                        Thread.sleep(1000)
                    } else {
                        break
                    }
                }
                stop()
            }
        }
    }

    override fun stop() {
        bar.isVisible = false
        bar.removeAll()
    }

    override fun setTime(seconds: Float) {
        this.seconds = seconds
    }

    override fun setTitle(title: String) {
        bar.setTitle(title)
    }
}